### PR TITLE
Add extras for fastparquet and pyarrow

### PR DESCRIPTION
**Pull request by crepererum**
This is helpful to ensure that the correct version of these libs is
installed.

What's the minimal version of `fastparquet` that dask supports?

### CODE REVIEW

**Comment by martindurant** [comment]
I think this would be better-covered in the functions that actually call the external packages, so that it would work for whichever install/packing solution people are using

**Comment by crepererum** [comment]
> I think this would be better-covered in the functions that actually call the external packages, so that it would work for whichever install/packing solution people are using

The issue is that not all versions work. E.g. `pyarrow` requires at least version `0.13.1`. Especially when using tools like [`pip-compile`](https://github.com/jazzband/pip-tools/), I expect dask to report the proper version constraints. That's exactly the point of declaring extra dependencies.

**Comment by martindurant** [comment]
The trouble, though, it that most people use `dask[complete]`, which will then install all of these things. The definition of "complete" could be changed too, of course.

**Comment by crepererum** [comment]
> The trouble, though, it that most people use dask[complete], which will then install all of these things. The definition of "complete" could be changed too, of course.

Sure, but why not include these two libs into `complete`?

**Comment by martindurant** [comment]
There are *lots* of things we could include in `complete` but don't, and in this particular instance, it is a near certainty that a user doesn't need both libs

**Comment by crepererum** [comment]
So how about adding these 2 extras but NOT adding them to `complete`?

**Comment by martindurant** [comment]
That sounds much more reasonable to me.
@jcrist @jakirkham , any thoughts?

**Comment by jcrist** [comment]
I'm against adding these to `"complete"`, and am -0.5 to adding them to extras at all. Generally we recommend users strictly specify their dependencies themselves, and only provide extras for specific sub packages (`array`, `dataframe`, etc...). Adding extras for a specific sub-library seems odd to me (why not specify that sub-library yourself in your requirements file?). It also opens the door to us having a boundless set of extras, since we have many many optional dependencies.

Also, historically not all versions of arrow have been supported (e.g. recently arrow 0.13.0 broke for us, and had to wait until 0.13.1 for a fix). Released versions of dask would have had e.g `arrow >= 0.12.0` for an arrow "extra" dependency, which would no longer be valid given this failure - users would still need to specify a version of arrow in their requirements file themselves.

I guess I don't see the point of adding an `extra` for only a single library dependency. And it requires users to specify `dask[arrow]` when installing to get that requirement, otherwise `pip install dask arrow==0.12.0` would still work fine.

Can you point to other projects that include extras for all optional dependencies?

**Comment by jakirkham** [comment]
Maybe we could call it `io`? Then we could add other related things.

If the version constraints are more complicated than `pip install` or `conda install` of some package (like additional version requirements) that seems like a good argument for us listing that dependency somewhere with that constraint. Agree/disagree? Thoughts?

**Comment by crepererum** [comment]
So let me explain the current workflow that someone has to use to work with dask:

You install `dask[whatever]`, but some features are missing (e.g. parquet reading). To figure out which package and which version is required for that, you have to either do a trail-run and hope that the exception message is explanatory enough or you have to look for the source code. None of these I find particularly user-friendly. Now if you update dask (and this can be a minor version!) you have to do this dance again.

For production usage however, you also want to use either `conda` or `pip-compile` (or any other solver) to make sure all packages work together, because pip just emits a warning but breaks things anyway. To feed constraints into the solver, they actually need to be specified, which is currently not done anyway in a machine-readable way. 

Apart from production usage, this also messes up (integration) tests with other packages, see JDASoftwareGroup/kartothek#111 for example.

Prior Art:
- [Airflow](https://github.com/apache/airflow/blob/61595040d48aa8873216c5f2bdd037ca82da4a2f/setup.py#L144-L285)
- [setuptools](https://setuptools.readthedocs.io/en/latest/setuptools.html?highlight=declaring-extras-optional-features-with-their-own-dependencies#declaring-extras-optional-features-with-their-own-dependencies)

> Maybe we could call it io? Then we could add other related things.

Sounds great.

Tl;Dr; It's nice for users.

**Comment by jakirkham** [comment]
Near term you may consider sending a PR to add `pyarrow` and `parquet` to [the `dask` feedstock]( https://github.com/conda-forge/dask-feedstock/blob/master/recipe/meta.yaml ). That's more or less a [kitchen-sink]( https://www.dictionary.com/browse/kitchen-sink ) install. So adding a few more things there is not a big deal.

**Comment by crepererum** [comment]
> Near term you may consider sending a PR to add pyarrow and parquet to the dask feedstock. That's more or less a kitchen-sink install. So adding a few more things there is not a big deal.

I think we could hopefully agree on both, the wheel/sdist extras and the feedstock. I can take care of the feedstock PR.

**Comment by mrocklin** [comment]
> You install dask[whatever], but some features are missing

The challenge here is that because Dask is used with dozens of other libraries we're probably not capable of keeping a comprehensive list of libraries up to date.

> Near term you may consider sending a PR to add pyarrow and parquet to the dask feedstock. That's more or less a kitchen-sink install. So adding a few more things there is not a big deal.

I'm actually somewhat against this, for the reasons mentioned above.  There are hundreds of libraries that we could place in the `dask` feedstock.  If we add all of them then people won't use it.

**Comment by martindurant** [comment]
(pyarrow, in particular, is a very large thing to install)

**Comment by jcrist** [comment]
> Maybe we could call it io? Then we could add other related things.

Users are unlikely to use both pyarrow and fastparquet, so sticking them both in the same extra seems counterproductive.

> If the version constraints are more complicated than pip install or conda install of some package (like additional version requirements) that seems like a good argument for us listing that dependency somewhere with that constraint. 

I think this is our best option. Dask has too many optional dependencies for using extras to be feasible, and pip doesn't allow optional package constraints (I *think* conda does now?). A table with extra dependencies, what they enable, and acceptable versions should be fairly easy for us to put together and maintain. @crepererum, would this be acceptable for your use case?

**Comment by mrocklin** [comment]
> and acceptable versions should be fairly easy for us to put together and maintain

I disagree somewhat, or at least I disagree if you consider all libraries that we touch.  Who knows the right versions of itk or tpot or rasterio or MDAnalysis or scikit-image or geopandas that are good to use with Dask?  We've had to touch all of those at some point for better interoperability.  I think that if you open this door we'll be pestered fairly consistently to keep this list up to date. 

If other folks want to maintain this on their time that's fine with me, but I'd be sad to see discussions here distract from other things that maintainers could be doing.

**Comment by jcrist** [comment]
> Who knows the right versions of itk or tpot or rasterio or MDAnalysis or scikit-image or geopandas that are good to use with Dask?

Those dependencies are going the other way (they may use dask internally, dask doesn't use them internally). I mean *only* things that dask (the library) internally conditionally imports. I think this is a far more manageable list, and many of these have version checks in the code already for validating these issues. We just need to keep the checks and docs in sync.

**Comment by TomAugspurger** [comment]
FWIW, pandas faces the same issue with many optional dependencies. We recently cleaned up their handling in https://github.com/pandas-dev/pandas/pull/26802 and put together a nice table. Maintaining that hasn't been too burdensome yet.

https://pandas.pydata.org/pandas-docs/stable/install.html#optional-dependencies

**Comment by crepererum** [comment]
> FWIW, pandas faces the same issue with many optional dependencies. We recently cleaned up their handling in pandas-dev/pandas#26802 and put together a nice table. Maintaining that hasn't been too burdensome yet.

The point is that the burden is now moved to the consumers, because there is not machine-readable standardized way of this.

> The challenge here is that because Dask is used with dozens of other libraries we're probably not capable of keeping a comprehensive list of libraries up to date.

I think this is a feature of a well maintained package, for the reasons outlined above.

**Comment by crepererum** [comment]
> I think this is a feature of a well maintained package, for the reasons outlined above.

Don't get me wrong. I think dask is a very good package and a nice maintainership. I just think on top of this, the full dependency declaration should be the default.

**Comment by TomAugspurger** [comment]
I wasn't aware that users could specify multiple extras like

```
# requirements.txt
dask[dataframe,pyarrow,fastparquet]
```

That increases the attractiveness of this to me, since then Dask is able to specify a minimum version of pyarrow / fast parquet that we require. Otherwise, I don't see an advantage of this over

```
# requirements.txt
dask[dataframe]
pyarrow
fastparquet
```

**Comment by martindurant** [comment]
> Otherwise, I don't see an advantage of this over

I think the original point was, that the user doesn't know which versions to put in the requirements, and would prefer the package authors to specify this.

**Comment by jcrist** [comment]
I've stated that I think this path leads to an explosion of `extra` options in our `setup.py` (we have many optional dependencies), and doesn't really solve the issue. Say we had an extra on a past release of dask:

```
...
"pyarrow": "pyarrow": ["pyarrow >=0.12.0"],
...
```

And users were installing dask with pyarrow support as:

```
pip install dask[pyarrow, dataframe]
```

Then pyarrow 0.13.0 is released, which broke our parquet handling with pyarrow in a non-patchable way. Now users install scripts don't work, but the dependency in the released library can't be fixed. We're not about to release a bugfix release with just a dependency change in the `setup.py`, as we may not be in a good place in our development cycle to do so (and historically we only release from master, rather than release branches).

A dependency table in our docs is easier to update asynchronously from the release, as other dependencies update.

That said, if other maintainers think this is a good idea, I'm not 100% against it.

**Comment by crepererum** [comment]
> I think the original point was, that the user doesn't know which versions to put in the requirements, and would prefer the package authors to specify this.

Exactly.

> Then pyarrow 0.13.0 is released, which broke our parquet handling with pyarrow in a non-patchable way. Now users install scripts don't work, but the dependency in the released library can't be fixed. We're not about to release a bugfix release with just a dependency change in the setup.py, as we may not be in a good place in our development cycle to do so (and historically we only release from master, rather than release branches).

That's a very good point. You are right that this is actually an issue if we don't want to be overly restrictive w/ our dependencies (e.g. only allowing things up to the last known version). I still think that the constraints could be helpful since they are at least better than nothing, but as you've pointed out, they may never be sufficient.

**Comment by TomAugspurger** [comment]
Is it fair to say that the only use case this helps out with is when a user has a pre-existing environment with, say, pyarrow=0.12, and then attempts to install dask into it? If the user could do `pip install dask[dataframe,pyarrow]` then they'd correctly get the pyarrow>=0.13. But if they `pip install dask[dataframe]` they'd be stuck with the wrong version of pyarrow.

Am I missing any other cases where it's helpful?

**Comment by crepererum** [comment]
> Am I missing any other cases where it's helpful?

One particular case is JDASoftwareGroup/kartothek#111 . At the time we had the following situation:
- kartothek blacklists pyarrow 0.14.X
- dask release hits pypi that requires at least pyarrow 0.13.1
- pyarrow never released 0.13.1 but only 0.13.0, 0.14.0 and 0.14.1

If the dependencies would have been declared in the suggested way, pip-compile would have already found this issue and either would have solved it during dependency resolution or would have produced a proper error message.

**Comment by TomAugspurger** [comment]
This has stalled a bit. Let's make a decision today, one way or the other.

To summarize

1. I think we're not considering adding these to `complete`. The choice is between a `fasparquet` and `pyarrow` extras, vs. better documentation of our optional dependencies.
2. The benefit to these extras is that downstream libraries *maybe* get a known compatible version of, e.g. pyarrow, when they `pip install dask[complete,pyarrow]`.

I say *maybe* because we aren't going to put a max pin on pyarrow, and when pyarrow makes an API breaking change we won't make a patch release just changing our setup.py. So the situation that it helps is when you have an existing env and are updating `dask`, but not explicitly updating pyarrow. Possibly it would help in https://github.com/dask/dask/pull/5209#issuecomment-519553186, but I don't understand what's going on there.

Is that a fair summarization?

Can we get a +/-1 (or 0 😄) from people (@jcrist @jakirkham @mrocklin)?

For my vote, I think I'm -0 on adding the extras. I'd prefer to focus specifying a minimum supported version in our docs and testing against that, and keeping that one place up to date.

**Comment by martindurant** [comment]
In the end, I think I'm -1: we can do this in docs. That would mean having a fairly easily-found page that we update as needed.

**Comment by jcrist** [comment]
I'm also -1, as stated above. We should add a table to our docs of optional dependencies. I don't think extras are the proper way to solve this problem.

**Comment by TomAugspurger** [comment]
Thanks.

Sorry @crepererum, but I think we'll pass on this PR. Thanks anyways.

I'll put up a PR later today starting a table of optional dependencies and their minimum versions (unless you want to @crepererum, but I completely understand if you're ready to be done with this issue 😄).
